Srecupe2 patch 1

srecupe2 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.


#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

srecupe2 committed:
Add files via upload

srecupe2 committed:
Add files via upload

srecupe2 committed:
Delete MultivariateFeatureSelector.py.html

srecupe2 committed:
Delete k_sample_test_final.py.html

srecupe2 committed:
Add files via upload

srecupe2 committed:
Add files via upload

srecupe2 committed:
Delete _multivariate_selection.py

srecupe2 committed:
Add files via upload

srecupe2 committed:
Add files via upload

srecupe2 committed:
Add files via upload

srecupe2 committed:
Add files via upload

srecupe2 committed:
Add files via upload

srecupe2 committed:
Add files via upload

srecupe2 committed:
Delete Test Class.py

srecupe2 committed:
Add files via upload

srecupe2 committed:
Add files via upload

srecupe2 committed:
Add files via upload

srecupe2 committed:
Add files via upload

srecupe2 committed:
Add files via upload

srecupe2 committed:
Add files via upload

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+     A k_sample test tests equality in distribution among groups. Groups
+    can be of different sizes, but must have the same dimensionality.
+    There are not many non-parametric k_sample tests, and this version
+    leverages the only k_sample multivariate independence test in scipy.
```
there is no "k_sample multivariate independence test in SciPy". This should say that this implementation reduces the k-sample testing problem to the independence testing problem, and leverages notable and powerful independence tests.

sampan501 replied:
cite the nonpar manova via independence testing paper

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/commit/0cca1d0f8c89244981a30c22928271880153b171

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+    y : ndarray of shape (n_samples,)
+        The target vector.
+    score_func : string that refers to a k_sample multivariate independence test from scipy
+                The default and only existing test is multiscale graph correlation.
```
Tabbing is off

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/commit/0cca1d0f8c89244981a30c22928271880153b171

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+    
+    Notes
+    -----
+    
```
No space

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/commit/0cca1d0f8c89244981a30c22928271880153b171

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+    Notes
+    -----
+    
+    1) The k_sample testing problem can be thought of as a generalization of
```
Bullets won't render properly on the website. See https://sublime-and-sphinx-guide.readthedocs.io/en/latest/lists.html

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/commit/0cca1d0f8c89244981a30c22928271880153b171
Revised structure to replicate other selectors

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+    concatenated versions of the inputs and a label matrix which are
+    paired. Then, any multivariate nonparametric test can be performed on
+    this data.
+    """
```
You never mentioned how we use it for multivariate feature selection

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/commit/0cca1d0f8c89244981a30c22928271880153b171

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+#!/usr/bin/env python
```
You not using PEP8 formatting, or whatever the formatting guidelines are in scikit-learn

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/pull/4/commits/184e5432f44d112873543cc2c7e0ae22e2ebb3f7

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+    matrices = []
+    for i in k_array:
+        indices = np.where(y == i)[0] 
+        if len(X.shape) == 1:
```
In the fit function, you should just use the `reshape` function instead of this if statement.

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/pull/4/commits/184e5432f44d112873543cc2c7e0ae22e2ebb3f7

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+    
+    #mgc case
+    if score_func == "mgc":
+        warnings.filterwarnings("ignore")
```
I think for now, just keep these 3 lines and remove the if statement. This can be modified later when Dcorr is added to SciPy

sampan501 replied:
Also, do something like
```
with warnings.catch_warnings():
     warnings.filterwarnings("ignore")
```

srecupe2 replied:
Do I need to put the mgc line inside of that statement

sampan501 replied:
yes inside the with. See https://github.com/neurodata/hyppo/blob/5db67a7473c0bc9009803f1a3b3143fdf92387de/hyppo/independence/mgc.py#L159 as a similar example

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/pull/4/commits/184e5432f44d112873543cc2c7e0ae22e2ebb3f7

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+######################################################################
+# Scoring function
+
+def k_sample_test(X, y,score_func="mgc"):
```
link back to hyppo somewhere in a comment

srecupe2 replied:
Is it ok if I say the function is a rewriting of hyppo.ksample.KSample
As in is that enough to say or should I also provide a link

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/commit/0cca1d0f8c89244981a30c22928271880153b171

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/pull/4/commits/184e5432f44d112873543cc2c7e0ae22e2ebb3f7

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+        warnings.filterwarnings("ignore")
+        mgc = multiscale_graphcorr(X,y,reps = 0)
+        stat = mgc.stat 
+    return(stat)
```
don't need parentheses

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/commit/0cca1d0f8c89244981a30c22928271880153b171

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+    def __init__(self, k=10):
+        self.k = k
+        
+    class _Parallel:
```
this doesn't have to be a nested class. I think it's confusing that way

sampan501 replied:
should have it's own docstring

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/commit/0cca1d0f8c89244981a30c22928271880153b171

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+
+        def __call__(self, index):
+            if np.var(self.X_new[:,index]) == 0:
+                stat = -1000.0
```
this should be 0

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/commit/0cca1d0f8c89244981a30c22928271880153b171

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+            if np.var(self.X_new[:,index]) == 0:
+                stat = -1000.0
+            else:   
+                if len(self.best_feat)==0:
```
spacing is off

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/pull/4/commits/184e5432f44d112873543cc2c7e0ae22e2ebb3f7

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+                stat = -1000.0
+            else:   
+                if len(self.best_feat)==0:
+                    X_j =  self.X_new[:,index] 
```
why don't you just reshape?

srecupe2 replied:
Im not sure exactly what you mean by reshape, this is basically my last edit left

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+                else:
+                    columns = self.best_feat.copy() 
+                    columns.append(index)
+                    X_j = self.X_new[:,columns]
```
these 2 lines are common between both conditions. So, this is more minimal:
```
if len(self.best_feat) != 0:
     columns = self.best_feat.copy()
     columns.append(index)
X_j = self.X_new[:,columns]
stat= k_sample_test(X_j,self.y)
```

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/commit/0cca1d0f8c89244981a30c22928271880153b171

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+        if isspmatrix(X) == True:
+            X = X.toarray()
+        
+        features = np.arange(X.shape[1])
```
comment this line

srecupe2 replied:
Which line specifically

sampan501 replied:
173

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+        
+        features = np.arange(X.shape[1])
+        
+        if np.isnan(X).any() == True:
```
These lines are too long according to PEP8 conventions. Reduce to 79 characters per line

sampan501 replied:
Instead of `==` write `is`

srecupe2 replied:
which line are you referring to?

sampan501 replied:
thie one (175)

srecupe2 replied:
nevermind i understand

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/commit/0cca1d0f8c89244981a30c22928271880153b171

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+        
+        if np.isnan(X).any() == True:
+            raise ValueError("existing non-parametric multivariate independence tests in scipy do not allow nan")
+        if type(self.k) != int:
```
Instead of `!=` write `is not`

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/commit/0cca1d0f8c89244981a30c22928271880153b171

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+            raise ValueError("existing non-parametric multivariate independence tests in scipy do not allow nan")
+        if type(self.k) != int:
+            raise ValueError("k features to select must be integer")
+        if not 0 <= self.k <= X.shape[1]:
```
Python chaining is kind of confusing, I would separate into two conditions and add an `and`

sampan501 replied:
also should be `0 <` not `0 <=`

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/commit/0cca1d0f8c89244981a30c22928271880153b171

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+        if not 0 <= self.k <= X.shape[1]:
+                raise ValueError("k features to select must be a non-negative integer that is less than or equal to n_features of X")
+        if not X.shape[0] >= 5:
+                raise ValueError("n_samples of data matrix X must be >= 5 in order to perform k_sample multivariate independence test")
```
rewrite as `number of samples is {}, must be >= 5` where `{}` is `X.shape[0]`

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/commit/0cca1d0f8c89244981a30c22928271880153b171

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+        if type(self.k) != int:
+            raise ValueError("k features to select must be integer")
+        if not 0 <= self.k <= X.shape[1]:
+                raise ValueError("k features to select must be a non-negative integer that is less than or equal to n_features of X")
```
rewrite as `k is {}, must be nonnegative <= number of features of X` where `{}` is `k`

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/commit/0cca1d0f8c89244981a30c22928271880153b171

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+        if np.isnan(X).any() == True:
+            raise ValueError("existing non-parametric multivariate independence tests in scipy do not allow nan")
+        if type(self.k) != int:
+            raise ValueError("k features to select must be integer")
```
This is a `TypeError`

sampan501 replied:
Rewrite as `k is type {}, must be int` where `{}` is `type(k)`

srecupe2 replied:
what exactly do you want in these brackets, because I have currently copied the brackets directly, but then as the ValueError occurs, brackets show but it seems like there should be user input inside, like it says float if its a float

sampan501 replied:
So the string should be something like `"{}".format(x)` where the value of `x` will show up inside the bracket

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/pull/4/commits/184e5432f44d112873543cc2c7e0ae22e2ebb3f7

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+            #of selected best features and each additional feature.
+            #size of operations in parallel per loop iteration is 
+            #n_features - len(best_features)
+            parallel = self._Parallel(X_new=X_new, y=y,best_features = best_features)
```
Since scikit-learn has joblib as a dependency, which in my opinion is better than multiprocessing. Do something like [this](https://github.com/neurodata/hyppo/blob/135cfcd6b13a986b09ec7629931bce8676ea5547/hyppo/tools/common.py#L456) which calls [this](https://github.com/neurodata/hyppo/blob/135cfcd6b13a986b09ec7629931bce8676ea5547/hyppo/tools/common.py#L368) for a single permutation

sampan501 replied:
space after `,` and remove spaces around `=`

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/commit/0cca1d0f8c89244981a30c22928271880153b171

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+            parallel = self._Parallel(X_new=X_new, y=y,best_features = best_features)
+            with MapWrapper(workers) as mapwrapper:
+                scores = list(mapwrapper(parallel, features)) 
+            scores_index = np.zeros((len(features),2)) 
```
197-199 can be condensed by concatenating `features` and `scores`

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/commit/0cca1d0f8c89244981a30c22928271880153b171

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+            scores_index[:,0] = features 
+            scores_index[:,1] = scores 
+            sorted_index = scores_index[scores_index[:, 1].argsort()] 
+            best = sorted_index[len(scores)-1,0] 
```
combine 201 and 202. Also isn't `best` already an int

srecupe2 replied:
Best has to be casted as it is float prior, if I copy what best is defined by in this line and use it later for the deletion, the line is too long, so I think we have to take it out on its own line

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+        #loop to select k best features, 
+        #each iteration adds next best feature
+        best_features = []
+        while len(best_features) < self.k: 
```
I would comment this loop a bit better

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/commit/0cca1d0f8c89244981a30c22928271880153b171

sampan501 commented on sklearn/feature_selection/tests/test_multivariate_select.py:
```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/env python
```
I think that these tests should just be added to the `test_feature_select.py` file

sampan501 replied:
add this line: https://github.com/srecupe2/scikit-learn/blob/e42fb227b891593520525cf6cb6302f0edfc12eb/sklearn/feature_selection/_univariate_selection.py#L1

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/pull/4/commits/184e5432f44d112873543cc2c7e0ae22e2ebb3f7

sampan501 commented on sklearn/feature_selection/tests/test_multivariate_select.py:
```diff
@@ -0,0 +1,147 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# In[ ]:
```
remove these comments in all the files

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# In[15]:
```
remove these comments

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/pull/4/commits/184e5432f44d112873543cc2c7e0ae22e2ebb3f7

sampan501 commented on sklearn/feature_selection/tests/test_multivariate_select.py:
```diff
@@ -0,0 +1,147 @@
+        shuffle=False,
+        random_state=0,
+    )
+    with pytest.raises(ValueError):
```
when doing 3 tests like this and just modifying a single parameters, use `pytest.mark.parametrize`. Here is an [example](https://github.com/srecupe2/scikit-learn/blob/e42fb227b891593520525cf6cb6302f0edfc12eb/sklearn/feature_selection/tests/test_feature_select.py#L93)

srecupe2 replied:
Commenting to note this as the issue that I am referring to.

sampan501 replied:
just add a `pytest.mark.parametrize` where the parameters are `k` and `ValueError` and the other is `k` and `TypeError`

sampan501 commented on sklearn/feature_selection/_multivariate_selection.py:
```diff
@@ -0,0 +1,214 @@
+#!/usr/bin/env python
+# coding: utf-8
+
```
Add these lines: https://github.com/srecupe2/scikit-learn/blob/e42fb227b891593520525cf6cb6302f0edfc12eb/sklearn/feature_selection/_univariate_selection.py#L3. Add you and me

srecupe2 replied:
Do you mean add us as authors in similar line but in the _multivariate_selection.py file? Do i also have to include these other authors?

sampan501 replied:
Yes just add us

srecupe2 replied:
https://github.com/srecupe2/scikit-learn/commit/0cca1d0f8c89244981a30c22928271880153b171

sampan501 commented on sklearn/feature_selection/tests/test_multivariate_select.py:
```diff
@@ -0,0 +1,147 @@
+import numpy as np
+from scipy import stats, sparse
+import pytest
+from sklearn.feature_selection import MultivariateFeatureSelector,k_sample_test
```
add space after comma

sampan501 reviewed (changes requested):
- You have to edit [`__init__.py`](https://github.com/srecupe2/scikit-learn/blob/srecupe2-patch-1/sklearn/feature_selection/__init__.py).
- Also, I don't think you modified the user guide?
- Look through [this](https://scikit-learn.org/stable/developers/contributing.html#how-to-contribute) and make sure you have followed everything.

Once you address this, let me know and I'll do a quick run-through again. After that, I think we can merge.

srecupe2 committed:
Add files via upload

fixed formatting for main py file, and removed the extra pynb comments, added author lines, linked back to hyppo, reworded k sample  test doc string as you requested, added references, added notes on how ksample used in multivariate feature selectaion, fixed tabbing, fixed space with Notes in Ksample, fixed parentheses, used joblib for parallel operation, wrote new nested function rather than nested class(no doc string followed from other selector class examples), fixed error messages and conditionals for error messages, concatenated features and scores,

srecupe2 committed:
Add files via upload

Fixed formatting and authors for real in this commit, only element not in the previous commit that I said was there

srecupe2 committed:
Add files via upload

fixed formatting, added authors,License, and quote line, took out the conditional in the fit function helper and kept in the ksample test matrix building, added warnings catch, took out the mgc if statement, commented that features line

srecupe2 committed:
Add files via upload

__init__.py modification

srecupe2 committed:
Add files via upload

Added pytest parametrization, added space, deleted comments at the top

srecupe2 committed:
Add files via upload

srecupe2 committed:
Add files via upload

Added import statements for MultivariateFeatureSelector and k_sample_test for the example plot file

srecupe2 committed:
Add files via upload

fixed bars for Multivariate Feature Selection

srecupe2 committed:
Update feature_selection.rst

Edit user guide
